src/rules/rule-builder.ts

```diff
@@ -0,0 +1,210 @@
+      logDebug(`Running ${rule.name}`);
+      return [rule.apply(text, options), true];
+    } else {
+      return [text, false];
+    }
+  }
+}
+
+export default abstract class RuleBuilder<TOptions extends Options> extends RuleBuilderBase {
+  abstract get OptionsClass(): (new() => TOptions);
+
+  static register<TOptions extends Options>(RuleBuilderClass: typeof RuleBuilderBase & (new() => RuleBuilder<TOptions>)): void {
+    const rule = RuleBuilderClass.getRule();
+    registerRule(rule);
+  }
+
+  safeApply(text: string, options?: Options): string {
+    return this.apply(text, this.buildRuleOptions(options));
+  }
+
+  buildRuleOptions(options?: Options): TOptions {
+    options = options ?? {};
+    const defaultOptions = new this.OptionsClass();
+    const ruleOptions = Object.assign(defaultOptions, options) as TOptions;
+
+    for (const optionBuilder of this.optionBuilders) {
+      optionBuilder.setRuleOption(ruleOptions, options);
+    }
+
+    return ruleOptions;
+  }
+
+  abstract get name(): string;
+  abstract get description(): string;
+  abstract get type(): RuleType;
+  abstract apply(text: string, options: TOptions): string;
+  abstract get exampleBuilders(): ExampleBuilder<TOptions>[];
+  abstract get optionBuilders(): OptionBuilderBase<TOptions>[];
+  get isSpecial(): boolean {
+    return false;
+  }
+
+  static applyIfEnabled<TOptions extends Options>(this: typeof RuleBuilderBase & (new() => RuleBuilder<TOptions>), text: string, settings: LinterSettings, extraOptions?: TOptions): [result: string, isEnabled: boolean] {
+    const rule = this.getRule();
+    return RuleBuilderBase.applyIfEnabledBase(rule, text, settings, extraOptions);
+  }
+
+  static getRuleOptions<TOptions extends Options>(this: (new() => RuleBuilder<TOptions>), settings: LinterSettings): TOptions {
+    const rule = RuleBuilderBase.getRule.bind(this)();
+    const builder = new this();
+    const optionsFromSettings = rule.getOptions(settings);
+    return builder.buildRuleOptions(optionsFromSettings);
+  }
+}
+
+export class ExampleBuilder<TOptions extends Options> {
+  readonly example: Example;
+
+  // HACK to bypass the Typescript generics system flaw
+  // https://github.com/microsoft/TypeScript/wiki/FAQ#why-is-astring-assignable-to-anumber-for-interface-at--
+  #_: TOptions;
+
+  constructor(args: {
+    description: string,
+    before: string,
+    after: string,
+    options?: TOptions
+  }) {
+    this.example = new Example(args.description, args.before, args.after, args.options);
+  }
+}
+
+type KeysOfObjectMatchingPropertyValueType<TObject, TValue> = {[TKey in keyof TObject]-?: TObject[TKey] extends TValue ? (TValue extends TObject[TKey] ? TKey : never) : never }[keyof TObject & string];
+
+type OptionBuilderConstructorArgs<TOptions extends Options, TValue> = {
+  OptionsClass: (new() => TOptions),
+  name: string
+  description: string,
+  optionsKey: KeysOfObjectMatchingPropertyValueType<TOptions, TValue>;
+};
```

Comment:
Does this give us any kind of protection against using bad property names or does this just help Typescript try to protect against bad property names?

Reply:
This is more for clarification on my end as I am not the most familiar with Typescript and tend to avoid generics unless it is absolutely necessary (I prefer inheritance to generics where possible).

Reply:
Looks like this allows us to get warning when a key is not present in the interface, so I will let my curiosity be satisfied on this one. It looks really nasty, but it seems to get the job done.